programs/helium-sub-daos/src/utils.rs

```diff
@@ -28,3 +30,96 @@ pub fn current_epoch(unix_timestamp: i64) -> u64 {
 pub fn next_epoch_ts(unix_timestamp: i64) -> u64 {
   (current_epoch(unix_timestamp) + 1) * u64::try_from(EPOCH_LENGTH).unwrap()
 }
+
+pub fn update_subdao_vehnt(sub_dao: &mut SubDaoV0, curr_ts: i64) {
+  let fall: u64 = PreciseNumber::new(sub_dao.vehnt_fall_rate.into())
+    .unwrap()
+    .checked_mul(
+      &PreciseNumber::new(
+        (curr_ts - sub_dao.vehnt_last_calculated_ts)
+          .try_into()
+          .unwrap(),
+      )
+      .unwrap(),
+    )
+    .unwrap()
+    .checked_div(&PreciseNumber::new(FALL_RATE_FACTOR).unwrap())
+    .unwrap()
+    .to_imprecise()
+    .unwrap()
+    .try_into()
+    .ok()
+    .unwrap();
+
+  sub_dao.vehnt_staked = sub_dao.vehnt_staked.checked_sub(fall).unwrap();
+  sub_dao.vehnt_last_calculated_ts = curr_ts;
+}
+
+pub fn calculate_voting_power(
+  d_entry: DepositEntry,
+  voting_mint_config: &VotingMintConfig,
+  amount_deposited_native: u64,
+  amount_initially_locked_native: u64,
+  curr_ts: i64,
+) -> Result<u64> {
+  if voting_mint_config.min_required_lockup_saturation_secs > 0
+    && d_entry.lockup.kind == LockupKind::None
+  {
+    return Ok(0);
+  }
+  let baseline_vote_weight = voting_mint_config.baseline_vote_weight(amount_deposited_native)?;
+
+  let min_required_locked_vote_weight =
+    voting_mint_config.min_required_lockup_vote_weight(amount_initially_locked_native)?;
+
+  let max_locked_vote_weight =
+    voting_mint_config.max_extra_lockup_vote_weight(amount_initially_locked_native)?;
+
+  let locked_vote_weight = d_entry.voting_power_locked(
+    curr_ts,
+    min_required_locked_vote_weight,
+    max_locked_vote_weight,
+    voting_mint_config.lockup_saturation_secs,
+    voting_mint_config.min_required_lockup_saturation_secs,
+  )?;
+
+  require_gte!(
+    max_locked_vote_weight,
+    locked_vote_weight,
+    ErrorCode::FailedVotingPowerCalculation,
+  );
+
+  baseline_vote_weight
+    .checked_add(locked_vote_weight)
+    .ok_or_else(|| error!(ErrorCode::FailedVotingPowerCalculation))
+}
```

Comment:
In the rewrite we have these, and `voting_mint_config.locked_vote_weight`
```rust
    locked_vote_weight_scaled_factor: u64,
    minimum_required_lockup_secs: u64,
    max_extra_lockup_vote_weight_scaled_factor: u64,
    genesis_vote_power_multiplier: u8,
    genesis_vote_power_multiplier_expiration_ts: i64,
```
Im going to be moving the changes over to this repo so take a look at those changes too